[1.21.4] Re-introduce IForgeItem.damageItem when an item takes damage

Jonathing opened the pull request:
`IForgeItem.damageItem()` allows the modder to override the amount of damage that would've been dealt to an item when it takes damage. This PR also fixes an issue where the player's `this.useItem` could be pre-emptively set to `ItemStack.EMPTY` and disallow any "broken item" variants.

- Fixes #10344 for 1.21.4

Jonathing committed:
Cache this.useItem before running the break logic

Jonathing committed:
Re-introduce IForgeItem.damageItem to work with fake breakage

I forgor to give the new method the callback function

Jonathing committed:
Add game test for fake shield that takes no damage

Jonathing commented:
Without patches:
![Screenshot_20250128_175611](https://github.com/user-attachments/assets/fca8b7e1-911f-43ac-95e6-9434e7bbe5bc)

With patches:
![Screenshot_20250128_175424](https://github.com/user-attachments/assets/41b9e5f4-c448-4180-8791-e2560b3a3f37)

LexManos commented on src/test/java/net/minecraftforge/debug/gameplay/item/PreventItemDamageTest.java:
```diff
@@ -0,0 +1,79 @@
+package net.minecraftforge.debug.gameplay.item;
+
```
License header

LexManos commented on src/test/java/net/minecraftforge/debug/gameplay/item/PreventItemDamageTest.java:
```diff
@@ -0,0 +1,79 @@
+
+        // shield on cooldown?
+        helper.assertTrue(initialDamage == shield.getDamageValue(), "Fake shield took damage! Expected: " + initialDamage + ", Actual: " + shield.getDamageValue());
+        helper.succeed();
```
You need to check that the itemstack in the main hand is still the shield.
Thats the whole 'bug'

Also, probably worth making the damage == 1 and make sure the current == inital -l so we know that the damage hook was called. Or add a Flag and have it flipped in damageItem

Jonathing committed:
Update licenses

Jonathing committed:
Finish game test, add more use cases and expand GameTestHelper

Jonathing committed:
Re-implement to be aware of hurt without break

Additionally, guard IForgeItem.damageItem behind guaranteed ignored damage. If the item stack is not damageable or the player has infinite materials (i.e. creative mode), the damage is never processed to begin with and will always return 0.

Jonathing commented:
Finally fully finished. Everything's been rebased and squashed into easily digestible commits.

Jonathing committed:
Add explanatory comments to new patches

Jonathing committed:
Reduce patch surface by removing unnecessary redundancies

Jonathing committed:
Add game test that directly tests FakeShieldTest.damageItem

Jonathing committed:
Fix shield-related tests not working in peaceful

Jonathing committed:
Gate stopping player item, ensure it is actually the use item

LexManos commented:
Just a note from our discussions, this is a bad name for this method. It doesn't actually damage the item, that is done by the applyDamage else ware. We'll have to figure out a better name/potential hook for those methods later.
